### PR TITLE
[FIX] account_*: payment method journal filter

### PR DESCRIPTION
**Pull request by vin-odoo**
Fix two issues linked to payment methods and their journal link.

SEPA Credit Transfer was marked as only available on EUR journals, but
from what we have been told, and we have seen, it should also be made
available for other currencies (CHF/SEK).
So we are changing the rule used to determine if SEPA Credit Transfer
is available on a journal to allow to use it on journals using CHF or
SEK as a currency.

There is another issue, where the filtering is done differently at the
journal creation and when the user add the payment method manually in
the lists.
The filter rules were not respected at the journal creation, which
would lead to incorrect default inbound and outbound payment method list.
For example, a new journal would have the company currency (let's say,
USD) but still have SEPA Credit Transfer (EUR,CHF,SEK) added on it by
default while it should not be available there.

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by qdp-odoo** [comment]
don't you need to add currency_id field also in the dependcies of `_compute_available_payment_method_ids`, for the same reason?

**Comment by qdp-odoo** [comment]
I don't understand the commit message use cases. Can you rephrase? or give an example?

**Comment by vin-odoo** [comment]
I don't think we need the currency_id on `_compute_available_payment_method_ids` as changing the currency will recompute de payment method and thus recompute the available payment methods.

**Comment by qdp-odoo** [comment]
@robodoo r+

File: addons/account/models/account_journal.py [hunk 1]
```diff
@@ -929,3 +929,21 @@ def _get_available_payment_method_lines(self, payment_type):
             return self.inbound_payment_method_line_ids
         else:
             return self.outbound_payment_method_line_ids
+
+    def _is_payment_method_available(self, payment_method_code):
+        """ Check if the payment method is available on this journal. """
+        self.ensure_one()
+        pm_info = self.env['account.payment.method']._get_payment_method_information().get(payment_method_code)
```
Comment (william-andre) [inline addons/account/models/account_journal.py:936]: Is this Java? :smile:

File: addons/account/models/account_journal.py [hunk 2]
```diff
@@ -929,3 +929,21 @@ def _get_available_payment_method_lines(self, payment_type):
...
+        if not pm_info:
+            return False
+        currency_ids = pm_info.get('currency_ids')
+        country_id = pm_info.get('country_id')
+        domain = pm_info.get('domain', [('type', 'in', ('bank', 'cash'))])
+        journal_types = domain[0][2]
```
Comment (william-andre) [inline addons/account/models/account_journal.py:942]: How do you know it's the first clause?
Still not perfect, but better:
```suggestion
        journal_types = next(right for left, operator, right in domain if left == 'type' and operator == 'in')
```

File: addons/account/models/account_journal.py [hunk 3]
```diff
@@ -929,3 +929,21 @@ def _get_available_payment_method_lines(self, payment_type):
...
+
+        is_available = (not currency_ids or
+                        ((self.currency_id.id is False and self.company_id.currency_id.id in currency_ids) or self.currency_id.id in currency_ids))
+        is_available = is_available and (not country_id or self.company_id.account_fiscal_country_id.id == country_id)
+        is_available = is_available and self.type in journal_types
```
Comment (william-andre) [inline addons/account/models/account_journal.py:947]:
```suggestion
        if (
            not currency_ids 
            or (self.currency_id.id is False and self.company_id.currency_id.id in currency_ids)  # check this, it doesn't make any sense
            or self.currency_id.id in currency_ids
        ):
            return False
        if country_id and self.company_id.account_fiscal_country_id.id != country_id:
            return False
        if self.type not in journal_types:
            return False
        return True
```

**Review by william-andre — CHANGES_REQUESTED** [review]

File: addons/account/models/account_journal.py [hunk 4]
```diff
@@ -929,3 +929,21 @@ def _get_available_payment_method_lines(self, payment_type):
...
+
+    def _is_payment_method_available(self, payment_method_code):
+        """ Check if the payment method is available on this journal. """
+        self.ensure_one()
+        pm_info = self.env['account.payment.method']._get_payment_method_information().get(payment_method_code)
+        if not pm_info:
+            return False
+        currency_ids = pm_info.get('currency_ids')
+        country_id = pm_info.get('country_id')
+        domain = pm_info.get('domain', [('type', 'in', ('bank', 'cash'))])
```
[the remark this answers is not in the document]
Comment (vin-odoo) [reply addons/account/models/account_journal.py:941]: Aaaaaaah I was wondering if something like this existed but wasn't sure. I'll check but this should be perfect for this situation thank you

**Review by FlorianGilbert — APPROVED** [review]
@robodoo r+

File: addons/account/models/account_journal.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (william-andre) [reply addons/account/models/account_journal.py:941]: So nothing?

[the remark this answers is not in the document]
Comment (vin-odoo) [reply addons/account/models/account_journal.py:941]: I can't really do it here without changing the method definition. And even then, it does a self.ensure_one() and I can't be sure we have a payment method for this code at this point in the db.
I will change this in master to use the method that returns the domain, just like you said but in stable I don't think it's possible.
